### PR TITLE
Nose-Hoover chain thermostated integrator.

### PR DESCRIPTION
**Pull request by andysim**
Reworked version of the [pull request to the master OpenMM](https://github.com/pandegroup/openmm/pull/1865) repository, to be more consistent with the existing integrators in this repo, particularly the Andersen thermostat.

### CODE REVIEW

**Comment by jchodera** [comment]
Awesome, thanks!

@andrrizzi @Lnaden : Can you review?

**Comment by andysim** [comment]
Has anybody seen [this](https://travis-ci.org/choderalab/openmmtools/jobs/263149741#L2840) problem with py36 before?  It appears to be unrelated to the changes I made, but I can't be absolutely certain.

**Comment by Lnaden** [comment]
Yeah, thats a test which stochastically fails on Travis-CI. I'll restart it

**Comment by Lnaden** [comment]
I should also be able to get through the review for this today

**Comment by andysim** [comment]
Excellent, thanks.  I don't actually have much experience with integrators and implemented this mostly for my own edification; if you have suggestions for default settings I'm more than happy to change them.  Just browsing the travis logs, it looks like the temp set/get is orders of magnitude slower than for the Andersen thermostat, which is strange.  The (probably too short to be significant) simulation timings look consistent between the different thermostats.  Regardless, I think the default collision frequency looks reasonable, but suspect the other parameters are probably overkill.  The `if` statements that codeclimate is complaining about are just to make the algorithm fall back to regular velocity verlet in the case that the chain length is zero.

File: openmmtools/integrators.py [hunk 1]
```diff
@@ -671,6 +672,206 @@ def __init__(self, temperature=298 * simtk.unit.kelvin, collision_rate=91.0 / si
...
+        self.addGlobalVariable("ndf", 0)      # number of degrees of freedom
+        self.addGlobalVariable("bathKE", 0.0) # Thermostat bath kinetic energy
+        self.addGlobalVariable("bathPE", 0.0) # Thermostat bath potential energy
+        self.addGlobalVariable("KE2", 0.0)    # Twice the kinetic energy
+        self.addGlobalVariable("Q", Q)        # Thermostat particle "mass"
+        self.addGlobalVariable("scale", 1.0)
+        self.addGlobalVariable("aa", 0.0)
+        self.addGlobalVariable("wdt", 0.0)
+        for w in range(self.n_ys):
+            self.addGlobalVariable("w%d"%w, self.weights[w])
```
Comment (Lnaden) [inline openmmtools/integrators.py]: Small consistency thing, can you switch all your string formatters to `.format` over `%`?

File: openmmtools/integrators.py [hunk 2]
```diff
@@ -671,6 +672,206 @@ def __init__(self, temperature=298 * simtk.unit.kelvin, collision_rate=91.0 / si
...
+    """
+
+
+    YSWeights = {
+        1 : [ 1.0000000000000000 ],
+        3 : [ 0.8289815435887510, -0.6579630871775020,  0.8289815435887510 ],
+        5 : [ 0.2967324292201065,  0.2967324292201065, -0.1869297168804260, 0.2967324292201065, 0.2967324292201065 ]
+    }
+
+    def __init__(self, temperature=298*simtk.unit.kelvin, collision_frequency=50/simtk.unit.picoseconds,
```
Comment (Lnaden) [inline openmmtools/integrators.py]: Can you change all the `simtk.unit` to just `unit`? I noticed at some point there was a double import at the top I can fix in a separate PR

Comment (andysim) [inline openmmtools/integrators.py]: I just removed the double import and all instances of `simkt.unit` in favor of `unit`, docstrings included.  If you didn't want me to make that many changes, please let me know and I can revert it (squashing the PR at the end is easy).

File: openmmtools/integrators.py (hunk 1 above)
Comment (andysim) [inline openmmtools/integrators.py]: Done!

**Comment by andysim** [comment]
Sorry, I forgot to port the test case over in the first version of the PR.  I just added a simple test that checks to make sure the total energy (system+bath) in conserved over 100 steps of dynamics.  I could make the test longer and also make sure the target temperature is achieved but, at ~10 seconds on my laptop, I figured this test is probably time consuming enough.

**Comment by jchodera** [comment]
Thanks, @andysim! 

We're tracking own some sort of segfault issue with the openmm `dev` package we use for testing. Hope to get this sorted out today.

**Comment by Lnaden** [comment]
> I just removed the double import and all instances of simkt.unit 

Thanks for doing that.

The rest of the code looks good though, I'll change to approval.

> tracking own some sort of segfault issue with the openmm 

I tracked down the issue and made changes to how our tests run in #275, this should stop the segfaults and allow the full battery to run. If those pass, we can probably get this merged today

**Comment by Lnaden** [comment]
@andysim Can you merge master into your fork'd branch here? I forgot how git works for a moment so this test will not pass until you merge the fix I did to the test into your PR

**Comment by andysim** [comment]
I'll `pull --rebase` and force a push back up to this branch.  Also, as a heads-up, I noticed that the doctest for ThermostatedIntegrator reports a FAIL in the Travis output.  That alone doesn't seem to constitute a test failure, but it's entirely possible that I caused a problem by removing the `simtk` from the example docstrings.

**Comment by Lnaden** [comment]
Thanks for seeing that. It appears to only happen in the dev version of the build, so I wonder if its related to the other failures. However, the tests in the main channel all pass, including the doctests, so I think its okay. If it crops up again, we'll handle it then.

**Pull request merged**